backend/remote-state/kubernetes/client.go

```diff
@@ -0,0 +1,342 @@
+			return "", err
+		}
+
+		// The secret doesn't exist yet, create it with the lock
+		sData := make(map[string][]byte)
+		sData[lockInfoKey] = lockInfo
+
+		label := map[string]string{
+			terraState:     "true",
+			terraKey:       c.nameSuffix,
+			terraWorkspace: c.workspace,
+		}
+		secret = &unstructured.Unstructured{
+			Object: map[string]interface{}{
+				"metadata": metav1.ObjectMeta{
+					Name:        sName,
+					Namespace:   c.namespace,
+					Labels:      label,
+					Annotations: map[string]string{"encoding": "gzip"},
+				},
+				"data": sData,
+			},
+		}
+
+		_, err = c.k8sClient.Create(secret, metav1.CreateOptions{})
+		if err != nil {
+			if !k8serrors.IsAlreadyExists(err) {
+				return "", err
+			}
+			// The secret was created between the get and create, grab it and keep going
+			secret, err = c.getSecret(sName)
+			if err != nil {
+				return "", err
+			}
+		} else {
+			// No error on the create so the lock has been created successfully
+			return info.ID, nil
+		}
+	}
+
+	li, err := c.getLockInfo(secret)
+	if err != nil {
+		return "", err
+	}
+
+	if li != nil {
+		// The lock already exists
+		lockErr := &state.LockError{
+			Info: li,
+			Err:  errors.New("lock already exists"),
+		}
+		return "", lockErr
+	}
+
+	setLockInfo(secret, lockInfo)
+
+	_, err = c.k8sClient.Update(secret, metav1.UpdateOptions{})
+	if err != nil {
+		return "", err
+	}
+
+	return info.ID, err
+}
+
+func (c *RemoteClient) Unlock(id string) error {
+	sName, err := c.createSecretName()
+	if err != nil {
+		return err
+	}
+
+	secret, err := c.getSecret(sName)
+	if err != nil {
+		// If the secret doesn't exist, there is nothing to unlock
+		if k8serrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+
+	lockInfo, err := c.getLockInfo(secret)
+	if err != nil {
+		return err
+	}
+
+	if lockInfo == nil {
+		// The lock doesn't exist
+		return nil
+	}
+
+	lockErr := &state.LockError{
+		Info: lockInfo,
+	}
+
+	if lockInfo.ID != id {
+		lockErr.Err = fmt.Errorf("lock id %q does not match existing lock", id)
+		return lockErr
+	}
+
+	setLockInfo(secret, []byte{})
+
+	_, err = c.k8sClient.Update(secret, metav1.UpdateOptions{})
+	if err != nil {
+		lockErr.Err = err
+		return lockErr
+	}
+
+	return nil
+}
+
+//getLockInfo takes a secret and attempts to read the lockInfo field.
+func (c *RemoteClient) getLockInfo(secret *unstructured.Unstructured) (*state.LockInfo, error) {
+	lockData, ok := getLockInfo(secret)
+	if len(lockData) == 0 || !ok {
+		return nil, nil
+	}
+
+	lockInfo := &state.LockInfo{}
+	err := json.Unmarshal(lockData, lockInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	return lockInfo, nil
+}
+
+func (c *RemoteClient) getSecret(name string) (*unstructured.Unstructured, error) {
+	return c.k8sClient.Get(name, metav1.GetOptions{})
+}
+
+func (c *RemoteClient) deleteSecret(name string) error {
+	delProp := metav1.DeletePropagationBackground
+	delOps := &metav1.DeleteOptions{PropagationPolicy: &delProp}
```

Comment:
It's a best-practice to also pass the [UID](https://github.com/kubernetes/apimachinery/blob/8f6ac2502e5143406f4eaedc4ca4262854d877ce/pkg/apis/meta/v1/types.go#L577) of the `Secret` in the `Preconditions` field, so you'll be 100% sure that you are deleting the same object.

For K8S >= 1.14 you can have the [ResourceVersion](https://github.com/kubernetes/apimachinery/blob/8f6ac2502e5143406f4eaedc4ca4262854d877ce/pkg/apis/meta/v1/types.go#L580).

Reply:
Normally, yes. But in this case I don't have a secret to start with, the call to this is creating the name then calling the delete.